rtrlib/aspa/aspa.c

```diff
@@ -176,368 +268,581 @@ static int compare_asns(const void *a, const void *b)
-			if (aspa_array_append(new_array, &current->record) != ASPA_SUCCESS) {
+			// Attempt to add record with $CAS twice.
+			// Error: Duplicate Add.
+			if (next_matches_current && next->type == ASPA_ADD) {
+				*failed_operation = next;
+				return ASPA_DUPLICATE_RECORD;
+			}
+
+			// This operation adds a record with $CAS, the next op however removes this $CAS record again.
+			// These form a no-op.
+			if (next_matches_current && next->type == ASPA_REMOVE) {
+#ifdef ASPA_NOTIFY_NO_OPS
+				// If it's a remove operation, we insert a reference to the removed record's providers.
+				next->record = current->record;
 #endif
+
+				// Mark as no-op.
+				current->is_no_op = true;
+				next->is_no_op = true;
+
+				// Skip next
+				i += 1;
+				continue;
+			}
+
+			// Add record by appending it to new array (copy providers)
+			if (aspa_array_append(new_array, &current->record, true) != ASPA_SUCCESS) {
 				*failed_operation = current;
 				return ASPA_ERROR;
 			}
-#ifdef ASPA_UPDATE_IN_PLACE
-			existing_i += 1;
-#endif
-			aspa_table_notify_clients(aspa_table, &current->record, rtr_socket, true);
 		}
-		// MARK: Removing a record
-		else if (op_type == ASPA_REMOVE) {
-#ifdef ASPA_UPDATE_IN_PLACE
-			if (!revert) {
-#endif
-				// Remember this provider set is now unused
-				if (append_to_set_array(unused_provider_sets, unused_size, &unused_capacity,
-							existing_record->provider_asns) != ASPA_SUCCESS) {
-					*failed_operation = current;
-					return ASPA_ERROR;
-				}
-#ifdef ASPA_UPDATE_IN_PLACE
-			}
-#endif
 
-			aspa_table_notify_clients(aspa_table, existing_record, rtr_socket, false);
-
-#ifdef ASPA_UPDATE_IN_PLACE
-			if (aspa_array_remove(array, existing_i) != ASPA_SUCCESS) {
+		// MARK: Handling 'remove' operations
+		else if (current->type == ASPA_REMOVE) {
+			// Attempt to remove record with $CAS, but record with $CAS does not exist
+			// Error: Removal of unknown record.
+			if (!existing_matches_current) {
 				*failed_operation = current;
-				return ASPA_ERROR;
+				return ASPA_RECORD_NOT_FOUND;
 			}
 
-			if (!revert) {
-				// Replace record in operation so it could be undone later.
-				current->record = *existing_record;
+			// Attempt to remove record with $CAS twice.
+			// Error: Removal of unknown record.
+			if (next_matches_current && next->type == ASPA_REMOVE) {
+				*failed_operation = next;
+				return ASPA_RECORD_NOT_FOUND;
 			}
-#endif
+
+			// "Remove" record by simply not appending it to the new array
+			existing_i += 1;
+
+			// If it's a remove operation, we insert a reference to the removed record's providers.
+			current->record.provider_count = existing_record->provider_count;
+			current->record.provider_asns = existing_record->provider_asns;
+		}
+	}
+
+	// Append remaining records (reuse existing provider array)
+	if (existing_i < array->size) {
+		for (; existing_i < array->size; existing_i++) {
+			aspa_array_append(new_array, aspa_array_get_record(array, existing_i), false);
 		}
 	}
```

Comment:
outer if statement is unnecessary
```suggestion
	for (; existing_i < array->size; existing_i++) {
		aspa_array_append(new_array, aspa_array_get_record(array, existing_i), false);
	}
```